### ISSUE TITLE
Check condition for initialising SentryNative

### ISSUE THREAD

**Issue by jamescrosswell** [getsentry/sentry-dotnet#3842]
Currently we have the following code that conditionally initialises the SentryNative SDK:

https://github.com/getsentry/sentry-dotnet/blob/55ea8c0e9258f8e239f5ed4da836ee7f40d23496/src/Sentry/SentrySdk.cs#L67-L72

The intent here is to enable the SentryNative SDK when C# applications are being compiled AOT to native code.

However currently we're not able to detect whether an application is being compiled AOT. See [this comment](https://github.com/dotnet/runtime/issues/94380#issuecomment-1794224768):
> There's no API to check for native AOT specifically. 

The check we are currently running probably more accurately determines whether the application has been Trimmed or not:
https://github.com/getsentry/sentry-dotnet/blob/cb0a35cc8550ab3eaffabea1eeac022a4d3f16b9/src/Sentry/Internal/AotHelper.cs#L18-L23

So it's possible we would be initialising SentryNative in managed applications at the moment (if trimming is performed during compilation/linking).

**Comment by phunkeler** [comment]
Would a preprocessor symbol (_defined somewhere in [Sentry.targets](https://github.com/getsentry/sentry-dotnet/blob/cb0a35cc8550ab3eaffabea1eeac022a4d3f16b9/src/Sentry/buildTransitive/Sentry.targets#L140C3-L159C19)_) work here?

**Comment by jamescrosswell** [comment]
> Would a preprocessor symbol (_defined somewhere in [Sentry.targets](https://github.com/getsentry/sentry-dotnet/blob/cb0a35cc8550ab3eaffabea1eeac022a4d3f16b9/src/Sentry/buildTransitive/Sentry.targets#L140C3-L159C19)_) work here?

Unfortunately not. The code that is initialising SentryNative is compiled into the Sentry code package... so compilation happens before the SDK user compiles their application (so it's not possible to know whether they're be AOT compiling their application to native code). 

Possibly we could use Sentry.targets to set an assembly attribute on the main assembly though and we could change the logic that initialises SentryNative to be conditional upon the presence of that assembly attribute rather than  trying to detect how the app was compiled/linked at runtime.

Alternatively we could leave the decision about whether or not to initialise SentryNative to the SDK user (e.g. pass it via the SentryOptions - default to `false`, of course). 

@vaind any thoughts?

**Comment by vaind** [comment]
What do we gain/loose by running it on managed apps with trimming enabled? How are stack traces handled on those at the moment? Do we have a sample app for this scenario in the repo?

**Comment by jamescrosswell** [comment]
> What do we gain/loose by running it on managed apps with trimming enabled? 

Good question. 
- It **_might_** not have any impact at all
- It **_might_** result in a slightly larger binary if the .NET trimmer doesn't try to trim native code
- It **_might_** register signal handlers that would be triggered for exceptions in unsafe code blocks (e.g. some of the profiling stuff)?

We could test this by building a trimmed (but not AOT compiled) app, comparing the binary size with/without SentryNative included and maybe trying to throw an exception from some unsafe code when SentryNative is included.

> How are stack traces handled on those at the moment? Do we have a sample app for this scenario in the repo?

I don't think stack traces should be affected by trimming. Implicitly, any code that gets trimmed, would never appear in a stack trace anyway.

**Comment by jamescrosswell** [comment]
> AFAIU, trimming is removing debug information, it won't affect what code gets executed (with the exception of code relying on the debug information at runtime, such as reflection).

Not quite. Whether debug symbols are included or stripped is separate from whether the code gets trimmed. 

Trimming reduces the application size by removing unused code. The trimming tool analyzes the application's dependencies and removes any libraries or code paths that are not referenced. That's why reflection can mess with it - if you've got a bit of code that takes an arbitrary string (e.g. some keyboard input) and uses that to create an instance of some class, it's very difficult (sometimes impossible) for the trimmer to work out which code needs to be retained to ensure that code will work at runtime. 

Serialisers (or more accurately deserialisers) have historically heavily relied on reflection to take arbitrary string values and use these to instantiate the C# types that those strings represent... so stuff like Newtsonsoft JSON and FastSerializer don't really play well with trimmed apps.

The only practical solution that I'm aware of is to use source generators rather than reflection to generate the code required to deserialize stuff. Source generators can't do everything that reflection can do. You have to know, at compile time, which types need to be deserialized... if you can't know that then you basically have to preserve every type that _could possibly_ be deserialized. If there are no constraints on which types might be deserialized, you have to preserve the entire .NET runtime plus every type in every third party library that the application references... so you end up with an app that's absolutely massive.

**Comment by whiskhub** [comment]
Please note it would be great if you could leave it up to the developer to enable Sentry Native. Mainly to allow catching native exceptions in WinUI (3) or any other C# apps that call into native C++ code that may throw exceptions. https://github.com/getsentry/sentry-dotnet/issues/2076#issuecomment-2440130964 https://github.com/getsentry/sentry-dotnet/issues/3712

**Comment by bruno-garcia** [comment]
> Mainly to allow catching native exceptions in WinUI (3) or any other C# apps that call into native C++ code that may throw exceptions.

When would you want to disable it? Right now it's always there when compiling with Native AOT.

**Comment by whiskhub** [comment]
I would want to _enable_ Sentry Native in JIT apps as well. But maybe that's a different topic/request.

**Comment by jamescrosswell** [comment]
> I would want to _enable_ Sentry Native in JIT apps as well. But maybe that's a different topic/request.

@whiskhub can JIT compiled WinUI 3 apps throw native exceptions? Do/could these come from the .NET Native libraries that WinUI depends on?

**Comment by whiskhub** [comment]
@jamescrosswell WinUI 3 does _not_ use .NET Native. WinUI 3 apps are normal Win32 apps. They use the Windows App SDK which contains the WinUI 3 C++ code, exposed as WinRT ("COM") APIs. WinUI 3 exceptions are usually mapped to C# exceptions, but not always (mainly not in "fatal" cases). Also, fail fast exceptions cannot be catched in C#.

But do not just think about WinUI 3 apps - any .NET JIT app that uses a C++ library with PInvoke can throw native exceptions. My current workaround is to enable both sentry-dotnet in C# and sentry-native in C++ (while I had to patch the sentry-native crashpad handler to also catch access violation exceptions).

Please also see my comment https://github.com/getsentry/sentry-dotnet/issues/2076#issuecomment-2440130964 for more details.

**Comment by bruno-garcia** [comment]
For native crash support in .NET on CoreCLR in JIT, lets move the convo here:
* https://github.com/getsentry/sentry-dotnet/issues/2076

We are not planning to tackle this right now because it's not just about adding `sentry-native`, but requires changes in the server to get proper C# <> C++ stack traces. Today Sentry's minidump processing is totally unaware of JIT'ed C# into native code.

**Comment by jamescrosswell** [comment]
> But do not just think about WinUI 3 apps - any .NET JIT app that uses a C++ library with PInvoke can throw native exceptions.

OK, that makes sense. 

The only decision to make then is whether to always initialise the Native SDK when initialising the .NET SDK or whether that should be optional (and, if so, what the default value should be). 

I'm thinking most .NET SDK users wouldn't be meddling with PInvoke and would get confused by a native stack trace (it doesn't relate to any of the code they've written) so perhaps we don't initialise the Native SDK by default but do allow people enable this via a flag in the options when initialising the .NET SDK.

@bruno-garcia @whiskhub thoughts?

**Comment by bruno-garcia** [comment]
For now lets only add sentry-native with Native AOT and hence only initialize sentry-native then. Lets track other things in other issues
